packages/token/src/erc721/extensions/erc721_enumerable/erc721_enumerable.cairo

```diff
@@ -0,0 +1,218 @@
+            self.ERC721Enumerable_all_tokens.read(index)
+        }
+
+        /// Returns the token id owned by `owner` at a given `index` of its token list.
+        /// Use along with `ERC721::balance_of` to enumerate all of `owner`'s tokens.
+        ///
+        /// Requirements:
+        ///
+        /// - `index` is less than `owner`'s token balance.
+        /// - `owner` is not the zero address.
+        fn token_of_owner_by_index(
+            self: @ComponentState<TContractState>, owner: ContractAddress, index: u256
+        ) -> u256 {
+            let erc721_component = get_dep_component!(self, ERC721);
+            assert(index < erc721_component.balance_of(owner), Errors::OUT_OF_BOUNDS_INDEX);
+            self.ERC721Enumerable_owned_tokens.read((owner, index))
+        }
+    }
+
+    //
+    // Internal
+    //
+
+    #[generate_trait]
+    pub impl InternalImpl<
+        TContractState,
+        +HasComponent<TContractState>,
+        impl ERC721: ERC721Component::HasComponent<TContractState>,
+        +ERC721Component::ERC721HooksTrait<TContractState>,
+        impl SRC5: SRC5Component::HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of InternalTrait<TContractState> {
+        /// Initializes the contract by declaring support for the `IERC721Enumerable`
+        /// interface id.
+        fn initializer(ref self: ComponentState<TContractState>) {
+            let mut src5_component = get_dep_component_mut!(ref self, SRC5);
+            src5_component.register_interface(interface::IERC721ENUMERABLE_ID);
+        }
+
+        /// Updates the ownership and token-tracking data structures.
+        ///
+        /// When a token is minted (or burned), `token_id` is added to (or removed from)
+        /// the token-tracking structures.
+        ///
+        /// When a token is transferred, minted, or burned, the ownership-tracking data structures
+        /// reflect the change in ownership of `token_id`.
+        ///
+        /// This must be added to the implementing contract's `ERC721HooksTrait::before_update`
+        /// hook.
+        fn before_update(
+            ref self: ComponentState<TContractState>, to: ContractAddress, token_id: u256
+        ) {
+            let erc721_component = get_dep_component!(@self, ERC721);
+            let previous_owner = erc721_component._owner_of(token_id);
+
+            if previous_owner.is_zero() {
+                self._add_token_to_all_tokens_enumeration(token_id);
+            } else if previous_owner != to {
+                self._remove_token_from_owner_enumeration(previous_owner, token_id);
+            }
+
+            if to.is_zero() {
+                self._remove_token_from_all_tokens_enumeration(token_id);
+            } else if previous_owner != to {
+                self._add_token_to_owner_enumeration(to, token_id);
+            }
+        }
+    }
+
+    #[generate_trait]
+    pub impl PrivateImpl<
```

Comment:
I vote we keep this as part of the InternalImpl and let the PrivateImpl refactor to a separate issue. The main reason I lean towards this is that it doesn't look right IMO to document the PrivateImpl in the docsite, because if it is private, why are we documenting it? But at the same time, it is not private, so we probably should document it. 

I think for the sake of not delaying this, we can stay with InternalImpl and design how PrivateImpl would look like in a different issue.

Reply:
Sounds good to me!